src/parse/parser.rs

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! A parser implementation.
+
+use super::{
+    extensions::{
+        simple_extension_declaration::{ExtensionFunction, ExtensionType, ExtensionTypeVariation},
+        simple_extension_uri::SimpleExtensionURI,
+    },
+    simple_extensions::{SimpleExtensionFunction, SimpleExtensions},
+    Anchor, Context, ContextError,
+};
+use crate::{
+    extensions::EXTENSIONS,
+    text::{
+        self,
+        simple_extensions::{SimpleExtensionsTypeVariationsItem, SimpleExtensionsTypesItem},
+    },
+};
+use std::{
+    borrow::Cow,
+    collections::{hash_map::Entry, HashMap},
+};
+use url::Url;
+
+/// A resolver for simple extensions.
+pub trait SimpleExtensionURIResolver: Default {
+    /// The error type of this resolver.
+    type Error: std::error::Error;
+
+    /// Attempt to resolve the given URI to the source text of a [SimpleExtensions].
+    fn resolve(&self, uri: &Url) -> Result<String, Self::Error>;
+}
+
+/// A parser implementation.
+#[derive(Default)]
+pub struct Parser<R: SimpleExtensionURIResolver = UreqResolver> {
+    extensions_uris: HashMap<Anchor<SimpleExtensionURI>, (SimpleExtensionURI, SimpleExtensions)>,
+    extension_types: HashMap<Anchor<ExtensionType>, (ExtensionType, SimpleExtensionsTypesItem)>,
+    extension_type_variations: HashMap<
+        Anchor<ExtensionTypeVariation>,
+        (ExtensionTypeVariation, SimpleExtensionsTypeVariationsItem),
+    >,
+    extension_functions:
+        HashMap<Anchor<ExtensionFunction>, (ExtensionFunction, SimpleExtensionFunction)>,
+    resolver: R,
+}
+
+#[cfg(feature = "parse-ureq")]
+/// A resolver that uses [ureq] to resolve simple extensions.
+pub struct UreqResolver {
+    agent: ureq::Agent,
+}
+
+impl Default for UreqResolver {
+    fn default() -> Self {
+        Self {
+            agent: ureq::Agent::new(),
+        }
+    }
+}
+
+impl SimpleExtensionURIResolver for UreqResolver {
+    type Error = ureq::Error;
+
+    fn resolve(&self, uri: &Url) -> Result<String, Self::Error> {
+        Ok(self.agent.get(uri.as_str()).call()?.into_string()?)
+    }
+}
+
+#[cfg(feature = "parse-reqwest")]
+#[derive(Default)]
+/// A resolver that uses [reqwest] to resolve simple extensions.
+pub struct ReqwestResolver {
```

Comment:
Oh, reqwest is a package name.  It may be useful to disable external checking of URIs -- that seems more like a validation step that might not be useful on systems just trying to use plans.